apps/ejabberd/src/odbc/ejabberd_odbc.erl

```diff
@@ -0,0 +1,484 @@
+            sql_query_internal([<<"rollback;">>], State),
+            {aborted, Reason};
+        {'EXIT', Reason} ->
+            %% Abort sql transaction on EXIT from outer txn only.
+            sql_query_internal([<<"rollback;">>], State),
+            {aborted, Reason};
+        Res ->
+            %% Commit successful outer txn
+            sql_query_internal([<<"commit;">>], State),
+            {atomic, Res}
+    end.
+
+-spec execute_bloc(fun(), state()) -> {'aborted', _} | {'atomic', _}.
+execute_bloc(F, _State) ->
+    case catch F() of
+        {aborted, Reason} ->
+            {aborted, Reason};
+        {'EXIT', Reason} ->
+            {aborted, Reason};
+        Res ->
+            {atomic, Res}
+    end.
+
+sql_query_internal(Query, #state{backend = Backend, db_ref = DBRef}) ->
+    case Backend:query(DBRef, Query) of
+        {error, "No SQL-driver information available."} ->
+            {updated, 0}; %% workaround for odbc bug
+        Result ->
+            Result
+    end.
+
+%% @doc Generate the OTP callback return tuple depending on the driver result.
+-spec abort_on_driver_error(_, state()) ->
+                                   {reply, Reply :: term(), state()} |
+                                   {stop, timeout | closed, state()}.
+abort_on_driver_error({error, "query timed out"} = Reply, State) ->
+    %% mysql driver error
+    {stop, timeout, Reply, State};
+abort_on_driver_error({error, "Failed sending data on socket" ++ _} = Reply, State) ->
+    %% mysql driver error
+    {stop, closed, Reply, State};
+abort_on_driver_error(Reply, State) ->
+    {reply, Reply, State}.
+
+
+-spec db_engine(Host :: odbc_server()) -> ejabberd_config:value().
+db_engine(Pool) when is_atom(Pool) ->
+    {ok, DbType, _} = wpool:call(Pool, get_db_info),
+    DbType;
+db_engine(Host) when is_binary(Host) ->
+    case ejabberd_config:get_local_option({odbc_server, Host}) of
+        SQLServer when is_list(SQLServer) ->
+            odbc;
+        Other when is_tuple(Other) ->
+            element(1, Other)
+    end.
+
+
+-spec connect(Backend :: module(), Settings :: term(), Retry :: non_neg_integer(),
+              RetryAfterSeconds :: non_neg_integer()) -> {ok, term()} | {error, any()}.
+connect(Backend, Settings, Retry, RetryAfterSeconds) ->
+    case Backend:connect(Settings) of
+        {ok, _} = Ok ->
+            Ok;
+        Error when Retry =:= 0 ->
+            Error;
+        Error ->
+            ?ERROR_MSG("Database connection attempt with ~p resulted in ~p."
+                       " Retrying in ~p seconds.", [Settings, Error, RetryAfterSeconds]),
+            timer:sleep(timer:seconds(RetryAfterSeconds)),
+            connect(Backend, Settings, Retry - 1, RetryAfterSeconds)
+    end.
+
+
+-spec backend(Host :: ejabberd:server()) -> module().
+backend(Host) when is_binary(Host) ->
+    Engine = atom_to_binary(db_engine(Host), latin1),
+    binary_to_atom(<<"ejabberd_odbc_", Engine/binary>>, latin1).
```

Comment:
In my humble opinion even if it wouldn't, we should find a way to make it work (e.g. defining somewhere list of possible engines).  Every *_to_atom with variable argument is a threat in some way.